File: source/com/barrybecker4/puzzle/tantrix1/solver/verification/InnerSpaceDetector.java
// Copyright by Barry G. Becker, 2012. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix1.solver.verification;

import com.barrybecker4.common.geometry.Box;
import com.barrybecker4.common.geometry.ByteLocation;
import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.puzzle.tantrix1.model.HexUtil;
import com.barrybecker4.puzzle.tantrix1.model.Tantrix;

import java.util.*;

import static com.barrybecker4.puzzle.tantrix1.model.HexTile.NUM_SIDES;

/**
 * Used to determine if a candidate solution has empty spaces within the tantrix.
 * A val;id solution cannot have such spaces.
 *
 *  @author Barry Becker
 */
public class InnerSpaceDetector {

    private Tantrix tantrix;

    /**
     * Constructor.
     * @param tantrix the tantrix state to test for solution.
     */
    public InnerSpaceDetector(Tantrix tantrix) {
        this.tantrix = tantrix;
    }

    /**
     * Start with an empty position on the border of the bbox.
     * Do a seed fill to visit all the spaces connected to that.
     * Finally, if there are any empty spaces inside the bbox that are not visited,
     * then there are inner spaces and it is not a valid solution.
     * @return true if there are no inner empty spaces.
     */
    public boolean hasInnerSpaces() {

        Set<Location> seedEmpties = findEmptyBorderPositions();
        Set<Location> visited = findConnectedEmpties(seedEmpties);
        return !allEmptiesVisited(visited);
    }

    /**
     * @return all the empty positions on the border
     */
    private Set<Location> findEmptyBorderPositions() {

        Box bbox = tantrix.getBoundingBox();
        Set<Location> empties = new HashSet<Location>();

        for (int i = bbox.getMinCol(); i <= bbox.getMaxCol(); i++) {
            Location loc = new ByteLocation(bbox.getMinRow(), i);
            if (tantrix.get(loc) == null)  {
                empties.add(loc);
            }
            loc = new ByteLocation(bbox.getMaxRow(), i);
            if (tantrix.get(loc) == null)  {
                empties.add(loc);
            }
        }

        for (int i = bbox.getMinRow() + 1; i < bbox.getMaxRow(); i++) {
            Location loc = new ByteLocation(i, bbox.getMinCol());
            if (tantrix.get(loc) == null)  {
                empties.add(loc);
            }
            loc = new ByteLocation(i, bbox.getMaxCol());
            if (tantrix.get(loc) == null)  {
                empties.add(loc);
            }
        }

        int totalLocs = (bbox.getHeight() + 1) * (bbox.getWidth() + 1);
        assert (totalLocs == tantrix.size() || empties.size() > 0):
                "We should have found at least one empty position on the border. Num Tiles ="
                + tantrix.size() + " bbox area = " + bbox.getArea();
        return empties;
    }

    private Set<Location> findConnectedEmpties(Set<Location> seedEmpties) {
        Set<Location> visited = new HashSet<Location>();

        Queue<Location> searchQueue = new LinkedList<Location>();
        searchQueue.addAll(seedEmpties);
        visited.addAll(seedEmpties);

        while (!searchQueue.isEmpty()) {
            Location loc = searchQueue.remove();
            List<Location> nbrEmpties = findEmptyNeighborLocations(loc);
            for (Location empty : nbrEmpties) {
                if (!visited.contains(empty)) {
                    visited.add(empty);
                    searchQueue.add(empty);
                }
            }
        }

        return visited;
    }

    /**
     * @return all the empty neighbor positions next to the current one.
     */
    private List<Location> findEmptyNeighborLocations(Location loc) {
        List<Location> emptyNbrLocations = new LinkedList<Location>();
        Box bbox = tantrix.getBoundingBox();

        for (byte i=0; i< NUM_SIDES; i++) {

            Location nbrLoc = HexUtil.getNeighborLocation(loc, i);
            if (tantrix.get(nbrLoc) == null && bbox.contains(nbrLoc)) {
                emptyNbrLocations.add(nbrLoc);
            }
        }
        return emptyNbrLocations;
    }

    /**
     * @param visited set of visited empties.
     * @return true if any empties in the tantrix bbox are not visited
     */
    private boolean allEmptiesVisited(Set<Location> visited) {
        Box bbox = tantrix.getBoundingBox();
        for (int i = bbox.getMinRow(); i < bbox.getMaxRow(); i++) {
            for (int j = bbox.getMinCol(); j <= bbox.getMaxCol(); j++)  {
                Location loc = new ByteLocation(i, j);
                if (tantrix.get(loc) == null && !visited.contains(loc))  {
                    return false;
                }
            }
        }
        return true;
    }

}

File: scala-source/com/barrybecker4/puzzle/tantrix/solver/verification/InnerSpaceDetector.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix.solver.verification

import com.barrybecker4.common.geometry.{Box, ByteLocation, Location}
import com.barrybecker4.puzzle.tantrix.model.{HexTile, HexUtil, Tantrix}

import scala.collection.mutable
import scala.collection.mutable.ListBuffer

/**
  * Used to determine if a candidate solution has empty spaces within the tantrix.
  * A valid solution cannot have such spaces.
  * @param tantrix the tantrix state to test for solution.
  * @author Barry Becker
  */
case class InnerSpaceDetector(tantrix: Tantrix) {

  /** Start with an empty position on the border of the bbox.
    * Do a seed fill to visit all the spaces connected to that.
    * Finally, if there are any empty spaces inside the bbox that are not visited,
    * then there are inner spaces and it is not a valid solution.
    * @return true if there are no inner empty spaces.
    */
  def numInnerSpaces(): Int = {
    val bbox = tantrix.getBoundingBox
    val seedEmpties = findEmptyBorderPositions(bbox)
    val visited = findConnectedEmpties(seedEmpties, bbox)
    numInternalEmpties(visited, bbox)
  }

  /**
    * Seeds for flood fill: empty cells that can reach the exterior of the bbox.
    * Includes axis-aligned rectangle edges (original) plus any empty hex in the bbox that has a
    * [[HexUtil]] neighbor strictly outside the bbox — needed on hex grids where every empty can lie
    * off the top/bottom/left/right lines of the bounding rectangle but still touch outside (otherwise
    * we would assert incorrectly and miss exterior-connected voids).
    */
  private def findEmptyBorderPositions(bbox: Box): Set[Location] = {
    val seeds = mutable.Set.empty[Location]
    for (i <- bbox.getMinCol to bbox.getMaxCol) {
      val top = new ByteLocation(bbox.getMinRow, i)
      if (tantrix(top).isEmpty) seeds += top
      val bottom = new ByteLocation(bbox.getMaxRow, i)
      if (tantrix(bottom).isEmpty) seeds += bottom
    }
    for (i <- bbox.getMinRow until bbox.getMaxRow) {
      val left = new ByteLocation(i, bbox.getMinCol)
      if (tantrix(left).isEmpty) seeds += left
      val right = new ByteLocation(i, bbox.getMaxCol)
      if (tantrix(right).isEmpty) seeds += right
    }
    for (r <- bbox.getMinRow to bbox.getMaxRow; c <- bbox.getMinCol to bbox.getMaxCol) {
      val loc = new ByteLocation(r, c)
      if (tantrix(loc).isEmpty) {
        var i = 0
        while (i < HexTile.NUM_SIDES) {
          val nbr = HexUtil.getNeighborLocation(loc, i)
          if (!bbox.contains(nbr)) seeds += loc
          i += 1
        }
      }
    }
    seeds.toSet
  }

  /** @return all empty region connected to a set of seed positions */
  private def findConnectedEmpties(seedEmpties: Set[Location], bbox: Box) = {
    var visited: Set[Location] = Set()
    val searchQueue: mutable.Queue[Location] = mutable.Queue()
    searchQueue ++= seedEmpties
    visited ++= seedEmpties
    while (searchQueue.nonEmpty) {
      val loc: Location = searchQueue.dequeue()
      val nbrEmpties = findEmptyNeighborLocations(loc, bbox)
      for (empty <- nbrEmpties) {
        if (!visited.contains(empty)) {
          visited += empty
          searchQueue.enqueue(empty)
        }
      }
    }
    visited
  }

  /** @return all the empty neighbor positions next to the current one. */
  private def findEmptyNeighborLocations(loc: Location, bbox: Box): List[Location] = {
    val buf = ListBuffer.empty[Location]
    for (i <- 0 until HexTile.NUM_SIDES) {
      val nbrLoc = HexUtil.getNeighborLocation(loc, i)
      if (tantrix(nbrLoc).isEmpty && bbox.contains(nbrLoc)) buf += nbrLoc
    }
    buf.toList
  }

  /** Count empty hexes inside the bbox that are not reached from border empty cells.
    * Rows run `minRow until maxRow` so the bottom bbox row is omitted here: those cells are on the border
    * and were seeded in [[findEmptyBorderPositions]], so any empty there is either a seed or was classified
    * via the side/corner border sweeps — not an enclosed cavity counted as "interior".
    *
    * @param visited set of visited empties
    */
  private def numInternalEmpties(visited: Set[Location], bbox: Box): Int = {
    val interiorRows = bbox.getMinRow until bbox.getMaxRow
    val interiorCols = bbox.getMinCol to bbox.getMaxCol
    var numEmpties = 0
    for (i <- interiorRows; j <- interiorCols) {
      val loc = new ByteLocation(i, j)
      if (tantrix(loc).isEmpty && !visited.contains(loc)) numEmpties += 1
    }
    numEmpties
  }
}
